compiler/luci/pass/src/FuseGRUPass.cpp

```diff
@@ -0,0 +1,674 @@
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FuseGRUPass.h"
+#include "helpers/NodeFiller.h"
+
+#include <luci/IR/CircleNodes.h>
+
+#include <luci/Profile/CircleNodeOrigin.h>
+#include <luci/Service/CircleNodeClone.h>
+
+#include <cmath>
+
+#include <cassert>
+
+// Helper to fuse GRU
+namespace
+{
+
+class GRUPatternBase
+{
+public:
+  GRUPatternBase(luci::CircleNode *candidate) { _pattern_last_node = candidate; }
+
+  virtual ~GRUPatternBase() = default;
+
+public:
+  virtual bool matched() = 0;
+
+public:
+  luci::CircleNode *_ifm = nullptr;
+  luci::CircleConst *_weight_ih = nullptr;
+  luci::CircleConst *_bias_ih = nullptr;
+  luci::CircleConst *_weight_hh = nullptr;
+  luci::CircleConst *_bias_hh = nullptr;
+
+  luci::CircleConst *_hidden_input = nullptr;
+
+  luci::CircleConst *_less_const = nullptr;
+
+  luci::CircleWhile *_while_node = nullptr;
+  luci::CircleWhileOut *_while_out_node = nullptr;
+  luci::CircleNode *_pattern_last_node = nullptr;
+};
+
+/**
+ * Below diagram shows GRU pattern to fuse.
+ * Note: this pattern for GRU with `return_sequences=False`
+ * - the below pattern will be replaced with one GRU
+ *  Main Graph:
+ *           [In]    [CircleConst]    [CircleConst]    [CircleConst]    [CircleConst]
+ *            |             |              |                |                 |
+ *            V             |              |                |                 |
+ *         [CircleWhile]<-----------------------------------------------------
+ *            |
+ *            V
+ *      [CircleWhileOut]
+ *            |
+ *            V
+ *          [Out]
+ *
+ * Condition Graph:
+ *          [In]   [CircleConst] (scalar int32 value)
+ *           |        |
+ *           V        |
+ *         [Less]------
+ *           |
+ *           V
+ *         [Out]
+ *
+ *  Body Graph must contain:
+ *  - 2 CircleFullyConnected nodes;
+ *  - 3 CircleMul nodes;
+ *  - 2 CircleLogistic nodes;
+ *  - 2 CircleSplit nodes;
+ *  - 6 CircleAdd nodes;
+ *  - 1 CircleGather node;
+ *  - 1 CircleReshape node;
+ *  - 1 CircleSub node;
+ *  - 1 CircleTanh node;
+ *  - 6 CircleSplitOut nodes;
+ *  - 5 CircleInput nodes;
+ *  - 5 CircleOutput nodes;
+ *
+ *  Body Graph:
+ *  [In_1]                     [In_2]--->[Add_2 (with Const)]--->[Out_2]            [In_3]
+ *    |    \                                                      |                   |
+ *    |     \                                         [In_4]---[Gather]      [Add_3 (with Const)]
+ *    |     [FullyConnected_1]                          |         |                   |
+ *    |               |                              [Out_4]      |                [Out_3]
+ *    |           [Split_1]                                 [FullyConnected_2]
+ *    |         /         |   \                                       |
+ *    |        |          |    \                                   [Split_2]
+ *    |    [Add_1] -------+----+---------------------------------/    |     |
```

Comment:
```suggestion
 *    |    [Add_1] ----------------------------------------------/    |     |
```

Reply:
Using `+` , I understand this as vertical line and horizontal lines are `connected`,
but I assume not.